tracer/src/Datadog.Trace/Tagging/TagsList.cs

```diff
@@ -3,389 +3,134 @@
-        protected void WriteMetric(ref byte[] bytes, ref int offset, string key, double value, ITagProcessor[] tagProcessors)
-        {
-            if (tagProcessors is not null)
-            {
-                for (var i = 0; i < tagProcessors.Length; i++)
-                {
-                    tagProcessors[i]?.ProcessMetric(ref key, ref value);
-                }
-            }
-
-            offset += MessagePackBinary.WriteString(ref bytes, offset, key);
-            offset += MessagePackBinary.WriteDouble(ref bytes, offset, value);
-        }
 
-        [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        protected void WriteMetric(ref byte[] bytes, ref int offset, byte[] keyBytes, double value, ITagProcessor[] tagProcessors)
-        {
-            if (tagProcessors is not null)
-            {
-                string key = null;
-                for (var i = 0; i < tagProcessors.Length; i++)
+                // If we get there, the tag wasn't in the collection
+                if (value != null)
                 {
-                    tagProcessors[i]?.ProcessMetric(ref key, ref value);
+                    tags.Add(new KeyValuePair<string, string>(key, value));
                 }
             }
-
-            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, keyBytes);
-            offset += MessagePackBinary.WriteDouble(ref bytes, offset, value);
         }
 
-        private int WriteMetrics(ref byte[] bytes, int offset, Span span, ITagProcessor[] tagProcessors)
+        public virtual void EnumerateTags<TProcessor>(ref TProcessor processor)
+            where TProcessor : struct, IItemProcessor<string>
         {
-            int originalOffset = offset;
-
-            // Start of "metrics" dictionary. Do not add any numeric tags before this line.
-            offset += MessagePackBinary.WriteStringBytes(ref bytes, offset, MetricsBytes);
-
-            int count = 0;
-
-            // We don't know the final count yet, write a fixed-size header and note the offset
-            var countOffset = offset;
-            offset += MessagePackBinary.WriteMapHeaderForceMap32Block(ref bytes, offset, 0);
-
-            // write "custom" span-level numeric tags (from list of KVPs)
-            var metrics = Metrics;
-            if (metrics != null)
+            var values = Volatile.Read(ref _tags);
+            if (values is not null)
             {
-                lock (metrics)
+                lock (values)
                 {
-                    count += metrics.Count;
-                    for (var i = 0; i < metrics.Count; i++)
+                    for (int i = 0; i < values.Count; i++)
                     {
-                        var metric = metrics[i];
-                        WriteMetric(ref bytes, ref offset, metric.Key, metric.Value, tagProcessors);
+                        processor.Process(new TagItem<string>(values[i].Key, values[i].Value, null));
```

Comment:
You create a `TagItem` here, a structure close to `KeyValuePair`. Why not switch the tags to a list of  `TagItem` instead in the first place?

Reply:
For an allocation reason: first that will force to change the `TagItem` from a `ref readonly struct` to a `readonly struct` because it will need to be allocated in the heap. The structure have another field (byte array) so we will be allocating more for each item in the list (the reference to the array), and that doesn't bring any benefit because that byte array will always be null for items in that List.